src/components/paster.ts

```diff
@@ -0,0 +1,230 @@
+import * as vscode from 'vscode'
+import * as path from 'path'
+import * as fs from 'fs'
+
+import { Extension } from '../main'
+
+export class Paster {
+    extension: Extension
+
+    constructor(extension: Extension) {
+        this.extension = extension
+    }
+
+    public async paste() {
+        this.extension.logger.addLogMessage('Performing formatted paste')
+
+        // get current edit file path
+        const editor = vscode.window.activeTextEditor
+        if (!editor) {
+            return
+        }
+
+        const fileUri = editor.document.uri
+        if (!fileUri) {
+            return
+        }
+        if (fileUri.scheme === 'untitled') {
+            vscode.window.showInformationMessage('Before paste image, you need to save current edit file first.')
+            return
+        }
+
+        const clipboardContents = await vscode.env.clipboard.readText()
+
+        if (clipboardContents === '') {
+            return
+        }
+
+        if (clipboardContents.split('\n').length === 1) {
+            const fpath = path.resolve(fileUri.fsPath, clipboardContents)
+            if (fs.existsSync(fpath)) {
+                this.pasteFile(editor, fileUri.fsPath, clipboardContents)
+                return
+            }
+        }
+        // if not pasting file
+        try {
+            this.pasteTable(editor, clipboardContents)
+        } catch (error) {
+            this.pasteNormal(editor, this.reformatText.completeReformat(clipboardContents))
+        }
+    }
+
+    public pasteNormal(editor: vscode.TextEditor, content: string) {
+        editor.edit(edit => {
+            const current = editor.selection
+
+            if (current.isEmpty) {
+                edit.insert(current.start, content)
+            } else {
+                edit.replace(current, content)
+            }
+        })
+    }
+
+    public pasteFile(editor: vscode.TextEditor, baseFile: string, file: string) {
+        const TABLE_FORMATS = ['.csv']
+        const extension = path.extname(file)
+
+        if (TABLE_FORMATS.indexOf(extension) !== -1) {
+            const contents = fs.readFileSync(path.resolve(baseFile, file), 'utf8')
+            if (extension === '.csv') {
+                // from: https://stackoverflow.com/a/41563966/3026698
```

Comment:
Hmm. Looking around I think https://www.npmjs.com/package/detect-tabular looks like it could be good (seems like a thin wrapper around three other libraries) however I haven't yet been able to work out how to use it…

Reply:
Why did you choose the package? Search the word, csv, not tabular.

Reply:
Ahh. I think that looks nice because it supports csv, json, xlsx, and ods.
Perhaps just use https://www.npmjs.com/package/csv-parser now and potentially that later.

Reply:
Now using csv-parser in 6b2d5471fe67dc920c24c9a91dfb3155ecef2c21